feat(curriculum): add contact form lab (#56270)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-contact-form/66f26c32ec6f90df01a44f60.md

---
id: 66f26c32ec6f90df01a44f60
title: Build a Contact Form
challengeType: 14
dashedName: lab-contact-form
demoType: onClick
---

# --description--

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

**User Stories:**

1. You should have a `div` with a class of `form-container` to hold your form.

2. You should have a `form` element within the `.form-container` `div`.

3. Within the `form` element, you should have the following elements and input fields:

   - A `h2` element with some text.
   - A text input field for the name with the `type` set to `text` and `id`, `name`, `required` attributes.
   - An email input field with the `type` set to `email` and `id`, `name`, `required` attributes.
   - A textarea for the message with `id`, `name`, and the `required` attribute.
   - A `button` element with the text `"Submit"` and a `type` attribute of `submit`.

> 4. Each `input` and `textarea` elements should have their corresponding `label` element with a `for` attribute set to the element's `id`.

Reviewer: I used a `input, textarea` selector in my CSS, which will not pass the tests for these. I think we should make a slight adjustment - maybe either mention that "You should have an `input` selector..." and "...`textarea` selector..." or adjust the tests to allow for more flexibility. You could check the computed values and not the stylesheet code.

Author: Added flexibility.

const labels = document.querySelectorAll('.form-container form label');

assert(labels.length > 0);

let found = false;

for (let label of labels) {
    if (label.getAttribute('for') === emailInput) {
        found = true;
        break;
    }
}

assert(found);
```

You should have a `label` element for the `#message` `textarea` with a `for` attribute matching the `id` value on the `textarea`.

```js
const messageInput = document.querySelector('.form-container form textarea')?.getAttribute('id');

const labels = document.querySelectorAll('.form-container form label');

assert(labels.length > 0);

let found = false;

for (let label of labels) {
    if (label.getAttribute('for') === messageInput) {
        found = true;
        break;
    }
}

assert(found);
```

The labels shouldn't be empty.

```js
const labels = document.querySelectorAll('.form-container form label');

assert(labels.length > 0);

for (let label of labels) {
    const labelText = label.textContent.trim();
    assert(labelText.length > 0);
}

```

The `form-container` `div` should have a background color.

```js
assert(new __helpers.CSSHelp(document).getStyle('.form-container')?.backgroundColor);
```

> The `form-container` `div` should have a visible border that is also rounded.

Reviewer: ```suggestion
The `form-container` `div` should have a `border-radius`.
```

Not sure on this one. It doesn't say anywhere to add a border. Maybe you meant that it should have a "visible border" in that, it should have a `background-color`, which makes it appear to have a border. We don't have to go with my suggestion.

Author: I added an outline as well, but that doesn't match the demo. So I have removed it.

assert(marginBottom);

```

Your `textarea` element should have a width, padding, margin on the bottom.

```js
const inputStyles = new __helpers.CSSHelp(document).getStyle('textarea');

const width = inputStyles?.width;
assert(width);

const padding = inputStyles?.padding;
const paddingTop = inputStyles?.paddingTop;
const paddingBottom = inputStyles?.paddingBottom;
const paddingLeft = inputStyles?.paddingLeft;
const paddingRight = inputStyles?.paddingRight;

assert(
  padding || (paddingTop && paddingBottom && paddingLeft && paddingRight)
);

const marginBottom = inputStyles?.marginBottom;
assert(marginBottom);
```

Your `button` element should have a background color.

```js
assert(new __helpers.CSSHelp(document).getStyle('button')?.backgroundColor);
```

Your `button` element should have a font size.

```js
assert(new __helpers.CSSHelp(document).getStyle('button')?.fontSize);
```

Your `button` element should have a hover effect that changes the background color.

```js
const aHoverColor = new __helpers.CSSHelp(document).getStyle('button:hover')?.getPropertyValue('color');
const buttonOriginalColor = new __helpers.CSSHelp(document).getStyle('button')?.getPropertyValue('color');

> assert(aHoverColor !== buttonOriginalColor);

Author: updated this one.

---
id: 66f26c32ec6f90df01a44f60
title: Design a Contact Form
challengeType: 14

> dashedName: design-a-contact-form

Reviewer 2: You shouldn't change this.

Changing the name to design was only for the title. the dashed name should stay the same as the others like this
` "dashedName": "lab-contact-form",`

Author: I have used the dashed name as `design-a-business-card` in the [business card lab](https://github.com/freeCodeCamp/freeCodeCamp/blob/main/curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md) as well.

The other projects are using `build-a-x` convention in the challenge file instead of `lab-x`

So should keep `design-a-contact-form` as it is or update the the business lab project as well along with contact form lab to `lab-x`?

Reviewer 2: oh I see, then we can leave it as is

4. Each `input` and `textarea` elements should have their corresponding `label` element with a `for` attribute set to the element's `id`.

5. The `.form-container` `div` should have a background color, and have values for `border-radius`, `padding`, and `width`.

6. The `label` elements should have a margin and a font color.

7. The `input` and `textarea` elements should have values for:
   
   - `width`.
   - `padding`.
   - Margin on the bottom.

8. The `button` element should have a:
   
   - Background color
   - Font size apart from the default.

9.  The button should have a hover effect that changes the background color.

**Note:** Be sure to link your stylesheet in your HTML and apply your CSS.

# --hints--

You should have a `div` with a class of `form-container`.

```js
const el = document.querySelector('div.form-container');
assert.isNotNull(el);
```

You should have a `form` element within the `.form-container` `div`.

```js
const el = document.querySelector('.form-container form');
assert.isNotNull(el);
```

Within the `form`, you should have an `h2` element with some text.

```js
const el = document.querySelector('.form-container form h2');

> assert(!!el && el.innerText.length > 0);

Reviewer 3: this also needs to be changed from `assert` to a specific method

Reviewer 3: ```suggestion
assert.isNotEmpty(el.innerText.length);
```

```

Within the `form`, you should have a text input field with the type set to `text`.

```js
const el = document.querySelector('.form-container form input[type="text"]');
assert.isNotNull(el);
```

Your `type=text` input field should have an `id`, a `name`, and a `required` attribute.

```js
const el = document.querySelector('.form-container form input[type="text"]');

assert(!!el && el.required);
assert.isNotEmpty(el?.getAttribute('id'));
assert.isNotEmpty(el?.getAttribute('name'));
```

Within the `form`, you should have an email input field with the `type` set to `email`.

```js
const el = document.querySelector('.form-container form input[type="email"]');
assert.isNotNull(el);
```

Your `type=email` input field should have an `id`, a `name`, and a `required` attribute.

```js
const el = document.querySelector('.form-container form input[type="email"]');

> assert(!!el && el.required);

Author: Updated this, it passed.

assert.isNotEmpty(el?.getAttribute('id'));
assert.isNotEmpty(el?.getAttribute('name'));
```

Within the `form`, you should have a `textarea` with an `id`, `name`, and a `required` attribute.

```js
const el = document.querySelector('.form-container form textarea');
assert.isNotNull(el);
assert.isNotNull(document.querySelector('.form-container form textarea[required]'));
assert.isNotEmpty(el?.getAttribute('id'));
assert.isNotEmpty(el?.getAttribute('name'));
```

Within the `form`, you should have a `button` element with a `type` attribute of `submit`.

```js
const el = document.querySelector('.form-container form button');

> assert(el);

Reviewer 3: ```suggestion
assert.exists(el);
```

Reviewer 3: or also just remove this line, this test is implicit in the `el?` in the next line

Author: Removed this.

assert.strictEqual(el?.getAttribute('type'), 'submit');
```

Your button should have a text of `"Submit"`.

```js
const el = document.querySelector('.form-container form button');
assert.strictEqual(document.querySelector('.form-container form button').textContent, 'Submit');
```

You should have a `label` element for the `#name` `input` with a `for` attribute matching the `id` value on the corresponding `input`.

```js
const nameInput = document.querySelector('.form-container form input[type="text"]')?.getAttribute('id');

const labels = document.querySelectorAll('.form-container form label');

assert.isAbove(labels.length, 0);

let found = false;

for (let label of labels) {
    if (label.getAttribute('for') === nameInput) {
        found = true;
        break;
    }
}

> assert(found);

Reviewer 3: ```suggestion
assert.isTrue(found);
```

```

You should have a `label` element for the `#email` `input` with a `for` attribute matching the `id` value on the corresponding `input`.

```js
const emailInput = document.querySelector('.form-container form input[type="email"]')?.getAttribute('id');

const labels = document.querySelectorAll('.form-container form label');

assert.isAbove(labels.length, 0);

let found = false;

for (let label of labels) {
    if (label.getAttribute('for') === emailInput) {
        found = true;
        break;
    }
}

assert(found);
```

You should have a `label` element for the `#message` `textarea` with a `for` attribute matching the `id` value on the `textarea`.

```js
const messageInput = document.querySelector('.form-container form textarea')?.getAttribute('id');

const labels = document.querySelectorAll('.form-container form label');

assert.isAbove(labels.length, 0);

let found = false;

for (let label of labels) {
    if (label.getAttribute('for') === messageInput) {
        found = true;
        break;
    }
}

> assert(found);

Reviewer 3: but, can't you check that `document.querySelector('label[for="${messageInput}"]')` exists?

> assert(found);

Author: `assert.exists(document.querySelector('label[for="${messageInput}"]'))` didn't pass for me.
I have left the test as it is for now.

> assert(found);

Reviewer 3: `label[for="${messageInput}"]'` needs to be in backticks for the string interpolation to work

> assert(found);

Author: updated the test.

```

The labels shouldn't be empty.

```js
const labels = document.querySelectorAll('.form-container form label');

assert.isAbove(labels.length, 0);

for (let label of labels) {
    const labelText = label.textContent.trim();

>     assert.isAbove(labelText.length, 0);

Reviewer 3: ```suggestion
    assert.isNotEmpty(labelText);
```

maybe this is more semantic?

}

```

The `form-container` `div` should have a background color.

```js

> assert(new __helpers.CSSHelp(document).getStyle('.form-container')?.backgroundColor);

Reviewer 3: ```suggestion
assert.isNotEmpty(new __helpers.CSSHelp(document).getStyle('.form-container')?.backgroundColor);
```

```

The `form-container` `div` should have a `border-radius`.

```js

> assert(new __helpers.CSSHelp(document).getStyle('.form-container')?.borderRadius);

Reviewer 3: ```suggestion
assert.isNotEmpty(new __helpers.CSSHelp(document).getStyle('.form-container')?.borderRadius);
```

```

The `form-container` `div` should have padding.

```js
assert(new __helpers.CSSHelp(document).getStyle('.form-container')?.padding);
```

The `form-container` `div` should have a width.

```js

> assert(new __helpers.CSSHelp(document).getStyle('.form-container')?.width);

Reviewer 3: ```suggestion
assert.isNotEmpty(new __helpers.CSSHelp(document).getStyle('.form-container')?.width);
```

```

Your labels should have a margin and font color.

```js
assert(new __helpers.CSSHelp(document).getStyle('label')?.margin);

> assert(new __helpers.CSSHelp(document).getStyle('label')?.color);

Reviewer 3: ```suggestion
assert.isNotEmpty(new __helpers.CSSHelp(document).getStyle('label')?.color);
```

```

> xxx Your `input` elements should have a width, padding, and a margin on the bottom.

Reviewer 2: looks like the xxx should be removed here

```js
// Helper function to extract properties from multiple selectors
function getStyleProperties(selectors, properties) {
  const result = {};
  properties.forEach((property) => {
    result[property] = selectors.map((selector) => selector?.[property]).find(Boolean);
  });
  return result;
}

// Initializing selectors
const selectors = [
  new __helpers.CSSHelp(document).getStyle('input'),
  new __helpers.CSSHelp(document).getStyle('input, textarea'),
  new __helpers.CSSHelp(document).getStyle('textarea, input')
];

// Extract width, padding, and margin properties
const { width } = getStyleProperties(selectors, ['width']);

> assert(width);

Reviewer 3: ```suggestion
assert.isNotEmpty(width);
```

const {
  padding,
  paddingTop,
  paddingBottom,
  paddingLeft,
  paddingRight
} = getStyleProperties(selectors, ['padding', 'paddingTop', 'paddingBottom', 'paddingLeft', 'paddingRight']);

// Assert padding properties

> assert(
>   padding ||
>   (paddingTop && paddingBottom && paddingLeft && paddingRight)
> );

Reviewer 3: ```suggestion
assert.isOk(
  padding ||
  (paddingTop && paddingBottom && paddingLeft && paddingRight)
);
```

for this one, I have no idea for a more specific method. `isOk` checks that the value is truthy, so it seems the more appropriate

Author: Updated this, it passed.

const { marginBottom } = getStyleProperties(selectors, ['marginBottom']);
assert(marginBottom);

```

Your `textarea` element should have a width, padding, margin on the bottom.

```js
// Helper function to extract properties from multiple selectors
function getStyleProperties(selectors, properties) {
  const result = {};
  properties.forEach((property) => {
    result[property] = selectors.map((selector) => selector?.[property]).find(Boolean);
  });
  return result;
}

// Initializing selectors
const selectors = [
  new __helpers.CSSHelp(document).getStyle('textarea'),
  new __helpers.CSSHelp(document).getStyle('input, textarea'),
  new __helpers.CSSHelp(document).getStyle('textarea, input')
];

// Extract width, padding, and margin properties
const { width } = getStyleProperties(selectors, ['width']);

> assert(width);

Reviewer 3: ```suggestion
assert.isNotEmpty(width);
```

const {
  padding,
  paddingTop,
  paddingBottom,
  paddingLeft,
  paddingRight
} = getStyleProperties(selectors, ['padding', 'paddingTop', 'paddingBottom', 'paddingLeft', 'paddingRight']);

// Assert padding properties

> assert(
>   padding ||
>   (paddingTop && paddingBottom && paddingLeft && paddingRight)
> );

Reviewer 3: ```suggestion
assert.isOk(
  padding ||
  (paddingTop && paddingBottom && paddingLeft && paddingRight)
);
```

Author: Updated this, it passed.

const { marginBottom } = getStyleProperties(selectors, ['marginBottom']);

> assert(marginBottom);

Reviewer 3: ```suggestion
assert.isNotEmpty(marginBottom);
```

```

Your `button` element should have a background color.

```js
assert(new __helpers.CSSHelp(document).getStyle('button')?.backgroundColor);
```

Your `button` element should have a font size.

```js
assert(new __helpers.CSSHelp(document).getStyle('button')?.fontSize);
```

Your `button` element should have a hover effect that changes the background color.

```js
const gs = (s) => new __helpers.CSSHelp(document).getStyle(s);
const btn_hover = gs('button:hover')?.backgroundColor;

> assert(btn_hover);

Reviewer 3: ```suggestion
assert.isNotEmpty(btn_hover);
```
